### PR TITLE
feat: added curved feature to axis component 

### PR DESCRIPTION
**Pull request by Moranski25**
Added feature to the axis component. The feature creates a curved axis with ticks and labels if the "isRadial"-property is set to true. The axis-properties that are new are listed below. 
arc:{
isRadial: boolean, (if not set defaults to false. Set to true for curved axis)
radius: number,  (if not set, defaults to 0.5)
startAngle: number,  (if not set, defaults to -pi/2)
endAngle: number,  (if not set, defaults to pi/2)
}
 
Skärmavbild 2024-11-06 kl  12 39 29
Skärmavbild 2024-11-04 kl  13 56 44
Skärmavbild 2024-11-04 kl  13 58 03

### CODE REVIEW

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js [hunk 1]
```diff
@@ -214,18 +230,30 @@ export default function nodeBuilder(isDiscrete) {
     const layered = state.labels.activeMode === 'layered';
     let majorTickNodes;
 
+    if (settings.isRadial) {
+      buildOpts.startAngle = settings.startAngle;
+      buildOpts.endAngle = settings.endAngle;
+      buildOpts.radius = settings.radius;
```
Comment (veinfors) [inline packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:236]: These new properties should be added in the specification, so they should be under the continous axis here:
https://github.com/qlik-oss/picasso.js/blob/4b1861a0c44c95a1d73b2d0d7f9ba9283ff32294/packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js#L210

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js [hunk 2]
```diff
@@ -214,18 +230,30 @@ export default function nodeBuilder(isDiscrete) {
...
+        nodes.push(buildLine(buildOpts));
+      }
     }
     if (settings.ticks.show) {
       buildOpts.style = settings.ticks;
       buildOpts.tickSize = settings.ticks.tickSize;
       buildOpts.padding = tickSpacing(settings);
-
-      majorTickNodes = tickBuilder(major, buildOpts);
+      buildOpts.tickColor = settings.ticks.tickColor || '#000';
```
Comment (veinfors) [inline packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:251]: Is this new tickColor setting necessary? 🤔
Where do current ticks (on a non-curved axis) get their color from?

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js [hunk 3]
```diff
@@ -214,18 +230,30 @@ export default function nodeBuilder(isDiscrete) {
...
+      if (settings.isRadial) {
+        majorTickNodes = arcTickBuilder(ticks, buildOpts);
```
Comment (veinfors) [inline packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:253]: Should we pass in `major` to `arcTickBuilder` as well maybe? The same as we pass in to `tickBuilder`...

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js (hunk 1 above)
Comment (Moranski25) [reply packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:236]: Yes. I am wondering if I should create a new object with these properties (buildOpts.radial.startAngle for example), or if I should add them as a property of the buildOpts-object (buildOpts.startAngle).

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js (hunk 2 above)
Comment (Moranski25) [reply packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js]: The tick stroke comes from settings.ticks.stroke, which can be specified when using picasso. So the new tickColor is not necessary! I fixed it in the last push

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js (hunk 1 above)
Comment (Moranski25) [reply packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js]: Fixed

File: packages/picasso.js/src/core/chart-components/axis/axis-arc-label-node.js [hunk 4]
```diff
@@ -86,5 +97,6 @@ export default function buildArcLabels(tick, buildOpts) {
   appendStyle(struct, buildOpts);
   appendBounds(struct, buildOpts);
   appendCollider(struct, tickPos);
+  struct.maxWidth = calculateMaxWidth(buildOpts, side, innerPos);
```
Comment (veinfors) [inline packages/picasso.js/src/core/chart-components/axis/axis-arc-label-node.js:100]: I think it's cleaner to have this on line 80 - `maxWidth: calculateMaxWidth(buildOpts, side, innerPos);`

File: packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js [hunk 5]
```diff
@@ -199,6 +199,23 @@ const DEFAULT_CONTINUOUS_SETTINGS = {
      * @type {boolean=} */
     show: true,
   },
+  /**
+   * @typedef {object}
+   */
+  arc: {
+    /** Flag that determines if axis is radial
+     * @type {boolean=} */
+    isRadial: false,
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js:208]: Should this be one more level up?

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js [hunk 6]
```diff
@@ -260,9 +300,11 @@ export default function nodeBuilder(isDiscrete) {
           tick,
         });
       };
-
       let labelNodes = [];
-      if (layered && (settings.align === 'top' || settings.align === 'bottom')) {
+      if (settings.arc.isRadial) {
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:304]: What happens if the settings does not have arc defined. The old versions does not have arc defined in settings.

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js [hunk 7]
```diff
@@ -214,18 +242,30 @@ export default function nodeBuilder(isDiscrete) {
...
+        nodes.push(buildLine(buildOpts));
+      }
     }
     if (settings.ticks.show) {
       buildOpts.style = settings.ticks;
       buildOpts.tickSize = settings.ticks.tickSize;
       buildOpts.padding = tickSpacing(settings);
-
-      majorTickNodes = tickBuilder(major, buildOpts);
+      if (settings.arc.isRadial) {
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:263]: What happens if the settings does not have arc defined. The old versions does not have arc defined in settings.

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js [hunk 8]
```diff
@@ -214,18 +242,30 @@ export default function nodeBuilder(isDiscrete) {
...
+      buildOpts.startAngle = settings.arc.startAngle;
+      buildOpts.endAngle = settings.arc.endAngle;
+      buildOpts.radius = settings.arc.radius;
+    }
     if (settings.line.show) {
       buildOpts.style = settings.line;
       buildOpts.padding = settings.paddingStart;
-
-      nodes.push(buildLine(buildOpts));
+      if (settings.arc.isRadial) {
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:253]: What happens if the settings does not have arc defined. The old versions does not have arc defined in settings.

File: packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js [hunk 9]
```diff
@@ -214,18 +242,30 @@ export default function nodeBuilder(isDiscrete) {
     const layered = state.labels.activeMode === 'layered';
     let majorTickNodes;
 
+    if (settings.arc.isRadial) {
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-node-builder.js:245]: What happens if the settings does not have arc defined. The old versions does not have arc defined in settings.

File: packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js (hunk 5 above)
Comment (veinfors) [reply packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js:208]: Yeah, I think I agree with that.
* arc should be optional as well.
* startAngle and endAngle should have default values specified if they are optional as well.

File: packages/picasso.js/src/core/chart-components/axis/axis-arc-label-node.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+function appendBounds(struct, buildOpts) {
+  struct.boundingRect = buildOpts.textBounds(struct);
+}
+
+export default function buildArcLabels(tick, buildOpts) {
+  const rect = buildOpts.innerRect;
+  const centerPoint = { cx: rect.width / 2, cy: rect.height / 2 };
+  const plotSize = Math.min(rect.height, rect.width) / 2;
+  const innerRadius = buildOpts.radius !== undefined ? plotSize * buildOpts.radius : plotSize * 0.5;
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-arc-label-node.js:50]: We can simplify it like this:
const innerRadius = plotSize * (buildOpts.radius ?? 0.5);

File: packages/picasso.js/src/core/chart-components/axis/axis-arc-node.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+function appendStyle(struct, buildOpts) {
+  extend(struct, buildOpts.style);
+}
+
+export default function buildArcLine(buildOpts) {
+  const rect = buildOpts.innerRect;
+  const centerPoint = { cx: rect.width / 2, cy: rect.height / 2 };
+  const plotSize = Math.min(rect.height, rect.width) / 2;
+
+  const innerRadius = plotSize * buildOpts.radius;
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-arc-node.js:12]: Should have the same formula as above:
const innerRadius = plotSize * (buildOpts.radius ?? 0.5);

File: packages/picasso.js/src/core/chart-components/axis/axis-arc-tick-node.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    x: centerX + radius * Math.cos(angle),
+    y: centerY + radius * Math.sin(angle),
+  };
+}
+
+export default function buildArcTicks(tick, buildOpts) {
+  const rect = buildOpts.innerRect;
+  const centerPoint = { cx: rect.width / 2, cy: rect.height / 2 };
+  const plotSize = Math.min(rect.height, rect.width) / 2;
+  const innerRadius = buildOpts.radius !== undefined ? plotSize * buildOpts.radius : plotSize * 0.5;
```
Comment (quanho) [inline packages/picasso.js/src/core/chart-components/axis/axis-arc-tick-node.js:18]: const innerRadius = plotSize * (buildOpts.radius ?? 0.5);

File: packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js (hunk 5 above)
Comment (Moranski25) [reply packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js:208]: I am thinking that "arc" should be on the same level as "line". 
Skärmavbild 2024-11-07 kl  10 39 01

Comment (veinfors) [reply packages/picasso.js/src/core/chart-components/axis/axis-default-settings.js:208]: Yeah, that I agree on, keep arc where you have it now, but move `isRadial` out of the `arc` object, one level up that is.

**Comment by bao-ho** [comment]
It seems that the space calculations for tick labels are stricter in the new chart than the old one. Look at 80,000,000 for example.

image

**Comment by bao-ho** [comment]
There seems to be a bug here that the 50,000,000.00 tick is smaller than the value (46,258,458.11)

image

But I think it's not related to this PR

File: junit.xml
Comment (bao-ho) [inline junit.xml:1]: I don't think this file is necessary. Probably add it to git ignore.

File: packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+      it('start on left side of tick', () => {
+        buildOpts.align = 'top';
+        tick = createTick(1, '0');
+        const result = buildArcLabels(tick, buildOpts);
+        expect(result.x).to.be.closeTo(40, 2);
+        expect(result.y).to.be.closeTo(305, 2);
+        expect(result.anchor).to.equal('end');
+      });
+
+      it('start on left side of tick', () => {
```
Comment (bao-ho) [inline packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js:76]: two tests should not have the same name (this and the one above it)

File: packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        buildOpts.align = 'top';
+        tick = createTick(0.7, '100');
+        const result = buildArcLabels(tick, buildOpts);
+        expect(result.x).to.be.closeTo(62, 2);
+        expect(result.y).to.be.closeTo(88, 2);
+        expect(result.anchor).to.equal('end');
+      });
+
+      it('start on right side of tick', () => {
```
Comment (bao-ho) [inline packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js:85]: fix the expectation

File: packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
+import { textBounds } from '../../../../web/text-manipulation';
+import buildArcLabels from '../axis-arc-label-node';
+
+function createTick(position, label) {
+  return {
+    position,
+    label,
+    value: 1.23, // just some dummy value to test node value for tracking animation
```
Comment (veinfors) [inline packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js:8]:
```suggestion
    value: 1.23,
```

File: packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-tick-node.spec.js [hunk 16]
```diff
@@ -0,0 +1,87 @@
+import buildArcTicks from '../axis-arc-tick-node';
+
+function createTick(position, value) {
+  return {
+    position,
+    value, // just some dummy value to test node value for tracking animation
```
Comment (veinfors) [inline packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-tick-node.spec.js:6]:
```suggestion
    value,
```

File: packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+  };
+}
+
+describe('Axis Label Node', () => {
```
Comment (veinfors) [inline packages/picasso.js/src/core/chart-components/axis/__tests__/axis-arc-label-node.spec.js:12]:
```suggestion
describe('Axis Arc Label Node', () => {
```

**Review by Moranski25 — COMMENTED** [review]
Feature that checks if a labels is out of bounds in y-direction.
Before: 

https://github.com/user-attachments/assets/4e9550c3-8331-4d91-930c-a5bf97aba806

After: 

https://github.com/user-attachments/assets/14f15c6d-70fc-49d3-8088-93993d7d5a94

**Review by quanho — APPROVED** [review]

**Pull request merged**